### ISSUE TITLE
Unclosed Span leads to OOM and breaks references

### ISSUE THREAD

**Issue by ledor473** [opentracing-contrib/java-spring-messaging#34]
Using the `OpenTracingChannelInterceptor` class may cause the application to go in OOM because the Span gets started but are not always closed. Also, since the Scope remains active, the references (wrong parent).

In the code below, the Span is created and is made active in the ScopeManager.
https://github.com/opentracing-contrib/java-spring-messaging/blob/1c5c7582d65f0ce99a52496dfad42fbefb3f5d56/opentracing-spring-messaging/src/main/java/io/opentracing/contrib/spring/integration/messaging/OpenTracingChannelInterceptor.java#L71-L72

It's supposed to be closed here: https://github.com/opentracing-contrib/java-spring-messaging/blob/1c5c7582d65f0ce99a52496dfad42fbefb3f5d56/opentracing-spring-messaging/src/main/java/io/opentracing/contrib/spring/integration/messaging/OpenTracingChannelInterceptor.java#L89-L93

That logic only works if the application is using a [`DirectChannel`](https://docs.spring.io/spring-integration/docs/5.0.5.RELEASE/reference/html/messaging-channels-section.html#channel-implementations-directchannel) but it's unlikely to work with something like the [`ExecutorChannel`](https://docs.spring.io/spring-integration/docs/5.0.5.RELEASE/reference/html/messaging-channels-section.html#executor-channel).
Furthermore, it appears that Spring Cloud Stream has a `DirectChannel` implementation that is not actually "direct" and needs to be excluded according to the [code in Sleuth](https://github.com/spring-cloud/spring-cloud-sleuth/commit/88ba09467cbe1fce19fbb515ca69be6a5d9e10e9).

Our issue was with the `DirectWithAttributesChannel` specifically, but the fix we are testing is to only activate the Span if it's a `DirectChannel`, pass both the Span and Scope in the headers and close both of them in the `afterSendCompletion`.

**Comment by pavolloffay** [comment]
thanks for reporting @ledor473, would you like to submit a PR to fix it?

**Comment by ledor473** [comment]
@pavolloffay yes we are still testing the change but I'll send something as soon as it's fixed
